Add documentation for pipeline templates (EDGE-963) (#2255)
buildkite/docs: pages/governance/pipeline_templates.md

# Pipeline templates

> 📘 Enterprise feature
> Pipeline templates is only available on an [Enterprise](https://buildkite.com/pricing) plan.

## Overview

Pipeline templates allow you to define standard pipeline step configurations to be used by all the pipelines in your organization.

> When a pipeline has a pipeline template assigned to it, the step configuration for that pipeline is inherited from the template and it is no longer possible to configure a custom step configuration for that pipeline via the UI.

Author: @mbelton-buildkite accepting this suggestion thanks - is "Buildkite dashboard" a shorthand for the whole UI that we use in the docs?

> When a pipeline has a pipeline template assigned to it, the step configuration for that pipeline is inherited from the template and it is no longer possible to configure a custom step configuration for that pipeline via the UI.

Reviewer: Yeah, it's a decision I inherited that I'm trying to go with. I'll probably revisit it sometime, but for now I want to see how well it fits in different places.

## Creating a pipeline template

Only administrators can create or update pipeline templates.

> 

Reviewer: Add a lead-in sentence for these steps:

```suggestion

To create a template:

```

> 1. Open your Buildkite _Organization Settings_ and choose [Templates](https://buildkite.com/organizations/-/pipeline-templates)

Reviewer: I think we should take them right there since we have a link:

```suggestion
1. Navigate to your [organization’s pipeline template settings](https://buildkite.com/organizations/-/pipeline-templates).
```

> 1. If this is your first template, click _Create a Template_. Otherwise choose _New Template_. A pipeline template is created automatically for you.

Reviewer: Click -> Select, missing comma:

```suggestion
1. If this is your first template, select _Create a Template_. Otherwise, select _New Template_. A pipeline template is created automatically for you.
```

Author: @mbelton-buildkite I'm not sure "select" is the right word here, but I can see the confusion.

The challenge I had with writing this instruction is that the page you see is entirely different depending upon whether you have created a template already or not.

So when you are yet to create a template you get:
![Empty state](https://github.com/buildkite/docs/assets/673828/2f17b47a-9275-4a4f-91e8-95c7be60304b)
Where as once you have a template it looks like:
![List view](https://github.com/buildkite/docs/assets/673828/aa6f6784-9966-44a4-acc5-b51da18c2033)

Perhaps it's easiest to just describe the empty state only?

Author: @mbelton-buildkite oh I just saw your next comment - I guess "select" is just the phrase you use for tap/click?

In that case, ignore this 😅

Reviewer: All good. I imagine most people are using Buildkite on a computer, but I think it's better to use the general term. This comes from [Microsoft's style guide](https://learn.microsoft.com/en-us/style-guide/a-z-word-list-term-collections/c/click).

> 1. Update the name and description of your new template and click _Apply_ to save your changes.

Reviewer: Click -> Select:

```suggestion
1. Update the name and description of your new template and select _Apply_ to save your changes.
```

> 1. You can also click _Edit_ to change the step configuration for the template.

Reviewer: Click -> Select:

```suggestion
1. You can also select _Edit_ to change the step configuration for the template.
```

> An administrator can add multiple templates to be used across the organization. Making changes and saving a template will apply those changes to all pipelines using that template.

Reviewer: Passive -> active voice:

```suggestion
An administrator can add multiple templates to use across the organization. Making changes and saving a template will apply those changes to all pipelines using that template.
```

## Testing a pipeline template

> At any time an administrator may test a pipeline template against a pipeline via the _New Build_ button on the pipeline page.

Reviewer: General tightening, avoid "via" because it can be ambiguous:

```suggestion
An administrator can test a pipeline template against a pipeline using the _New Build_ button on the pipeline page.
```

> If a template exists for the organization then it can be selected from the _Pipeline template_ dropdown to create a new build using the step configuration from that template.

Reviewer: Missing comma:

```suggestion
If a template exists for the organization, it can be selected from the _Pipeline template_ dropdown to create a new build using the step configuration from that template.
```
